### ISSUE TITLE
Impossible to stop Hazelcast Client service if it has never connected to the server.

### ISSUE THREAD

**Issue by simone** [hazelcast/hazelcast#10237]
Using service commons-daemon, in the stop() method.

Please help to me to understand how to force a shutdown of the connections and of the initialization.

After the shutdown of the service, this 2 threads are still running, no matter if you run shutdown, onShutdown, thread.interrupt(), thread.close(). 

       HazelcastInstanceFactory.terminateAll();
        hazelcasts.stream().map(instance -> (HazelcastClientInstanceImpl)instance).forEach(client -> {
            client.getConnectionManager().shutdown();
            for (ClientConnection connection : client.getConnectionManager().getActiveConnections()){
                connection.close("shutdown relay", new ShutdownChannelGroupException());
            }
            client.doShutdown();
        });
        hazelcasts.forEach(HazelcastInstance::shutdown);

Threads still up and running. 

```
2017-04-06 08:09:43,928 -  WARN -  - hz.Cluster.IO.thread-in-0 - com.hazelcast.client.connection.nio.ClientConnection - Cluster [CDDS Server] [3.8] ClientConnection{alive=false, connectionId=1, socketChannel=DefaultSocketChannelWrapper{socketChannel=java.nio.channels.SocketChannel[connected local=/172.31.1.8:57201 remote=/172.31.0.148:5701]}, remoteEndpoint=null, lastReadTime=2017-04-06 08:09:40.896, lastWriteTime=never, closedTime=2017-04-06 08:09:40.912, lastHeartbeatRequested=never, lastHeartbeatReceived=never, connected server version=null} lost. Reason: java.io.EOFException[Remote socket closed!]
2017-04-06 08:09:51,846 -  WARN -  - hazelcast-initialization-2-thread-1 - com.hazelcast.client.spi.impl.ClusterListenerSupport - Cluster [CDDS Server] [3.8] Unable to get alive cluster connection, try in 0 ms later, attempt 1 of 2147483647.
```

**Comment by pveentjer** [comment]
Which HZ version are you using?

**Comment by simone** [comment]
3.6 before, updated now to 3.8. same behaviour.

I'm investigating on NonBlockingIOThread that is not a daemon and does'n run on a pool.

**Comment by sancar** [comment]
Hi @simone , 
I am preparing a fix for 3.9 branch for this problem. But the fix was possible because we have async way of starting hazelcast client. 
In 3.9, you will be able to do following, 
```
 ClientConfig clientConfig = new ClientConfig();
        clientConfig.getConnectionStrategyConfig().setAsyncStart(true);
        clientConfig.getNetworkConfig()
                .setConnectionAttemptLimit(Integer.MAX_VALUE);
        HazelcastInstance client = HazelcastClient.newHazelcastClient(clientConfig);
        client.shutdown();
```

In 3.8, if you want client to shutdown sooner, only thing you can do is to set ConnectionAttemptLimit to a lower value.
